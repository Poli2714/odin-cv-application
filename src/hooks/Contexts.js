import { createContext } from 'react';

const IndexContext = createContext({
  activeIndex: null,
  setActiveIndex: () => {},
});

const KeyContext = createContext({
  activeKey: {
    eduActiveKey: null,
    expActiveKey: null,
  },
  setActiveKey: () => {},
});

const EduErrMsgContext = createContext({
  showEduErrMessage: false,
  setShowEduErrMessage: () => {},
});

const ExpErrMsgContext = createContext({
  showExpErrMessage: false,
  setShowExpErrMessage: () => {},
});

const PersonalInfoContext = createContext({
  personalInfo: {
    profilePicture: '',
    fullName: '',
    currentJobTitle: '',
    email: '',
    mobile: '',
    city: '',
  },
  setPersonalInfo: () => {},
});

const SummaryContext = createContext({
  summary: '',
  setSummary: () => {},
});

const ExperienceContext = createContext({
  experiences: {
    id: {
      jobTitle: '',
      employer: '',
      jobStartDate: '',
      jobEndDate: '',
      jobDescription: '',
    },
  },
  setExperiences: () => {},
});

const EducationContext = createContext({
  educations: {
    id: {
      degree: '',
      university: '',
      eduStartDate: '',
      eduEndDate: '',
      eduInfo: '',
    },
  },
  setEducations: () => {},
});

export {
  EducationContext,
  EduErrMsgContext,
  ExperienceContext,
  ExpErrMsgContext,
  IndexContext,
  KeyContext,
  PersonalInfoContext,
  SummaryContext,
};
