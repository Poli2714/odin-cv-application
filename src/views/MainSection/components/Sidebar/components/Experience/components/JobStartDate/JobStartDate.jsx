import { useContext } from 'react';
import PropTypes from 'prop-types';

import { InputWrapper } from 'src/components/layout';
import { Label, TextInput } from 'src/components/forms';
import { formatDateInput } from 'src/utils';
import { ExperienceContext } from 'src/hooks/Contexts';

function JobStartDate({ id }) {
  const { experiences, setExperiences } = useContext(ExperienceContext);
  const currentYear = new Date().getFullYear();
  const pattern = `^(?:0[1-9]|1[0-2]) / (?:19[0-9]{2}|20(?:00|[0-${currentYear - 1900}][0-9]))$`;

  const handleChangeJobStartDate = (e, key) => {
    const formattedDate = formatDateInput(
      e.target.value,
      experiences[key].jobStartDate,
    );
    if (formattedDate === experiences[key].jobStartDate) return;
    setExperiences({
      ...experiences,
      [key]: {
        ...experiences[key],
        jobStartDate: formattedDate,
      },
    });
  };

  return (
    <InputWrapper>
      <Label labelName="Start" labelFor="jobStartDate" />
      <TextInput
        name="jobStartDate"
        onChange={e => handleChangeJobStartDate(e, id)}
        pattern={pattern}
        placeholder="MM / YYYY"
        value={experiences[id].jobStartDate}
      />
    </InputWrapper>
  );
}

JobStartDate.propTypes = {
  id: PropTypes.string,
};

export default JobStartDate;
