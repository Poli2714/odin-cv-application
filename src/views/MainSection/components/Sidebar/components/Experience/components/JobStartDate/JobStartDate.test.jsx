import { expect, test, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useState } from 'react';
import PropTypes from 'prop-types';

import JobStartDate from './JobStartDate';
import { ExperienceContext } from 'src/hooks/Contexts';

vi.mock('src/components/layout', () => ({
  InputWrapper: vi.fn(({ children }) => (
    <div data-testid="input-wrapper">{children}</div>
  )),
}));

vi.mock('src/components/forms', () => ({
  Label: vi.fn(() => <label htmlFor="jobStartDate">Mock label</label>),
  TextInput: vi.fn(({ onChange, value }) => (
    <input type="text" id="jobStartDate" onChange={onChange} value={value} />
  )),
}));

function MockParentComponent({ id }) {
  const [experiences, setExperiences] = useState({
    testId: { jobStartDate: '' },
    testId2: { jobStartDate: '07 / 2017' },
  });

  return (
    <ExperienceContext.Provider value={{ experiences, setExperiences }}>
      <JobStartDate id={id} />
    </ExperienceContext.Provider>
  );
}

MockParentComponent.propTypes = {
  id: PropTypes.string,
};

test('renders JobStartDate', () => {
  render(<MockParentComponent id="testId" />);

  expect(screen.getByTestId('input-wrapper')).toMatchSnapshot();
  expect(screen.getByLabelText('Mock label')).toHaveValue('');
});

test('renders JobStartDate with different value', () => {
  render(<MockParentComponent id="testId2" />);

  expect(screen.getByLabelText('Mock label')).toHaveValue('07 / 2017');
});

test('updates jobStartDate value as a user types', async () => {
  const user = userEvent.setup();

  render(<MockParentComponent id="testId" />);
  const input = screen.getByLabelText('Mock label');

  await user.type(input, '022020');
  expect(input).toHaveValue('02 / 2020');

  await user.clear(input);
  await user.type(input, '101988');
  expect(input).toHaveValue('10 / 1988');

  await user.clear(input);
  await user.type(input, '07');
  expect(input).toHaveValue('07 / ');
});

test("doesn't update jobstartDate value as a user enters an invalid number", async () => {
  const user = userEvent.setup();

  render(<MockParentComponent id="testId" />);
  const input = screen.getByLabelText('Mock label');

  await user.type(input, '13');
  expect(input).toHaveValue('1');

  await user.type(input, '22024');
  expect(input).toHaveValue('12 / 202');
});
