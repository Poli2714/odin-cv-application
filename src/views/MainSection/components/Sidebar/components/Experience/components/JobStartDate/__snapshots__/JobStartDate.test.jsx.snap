// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders JobStartDate 1`] = `
<div
  data-testid="input-wrapper"
>
  <label
    for="jobStartDate"
  >
    Mock label
  </label>
  <input
    id="jobStartDate"
    type="text"
    value=""
  />
</div>
`;
